src/mtype.d

```diff
@@ -8748,29 +8754,34 @@ public:
         {
             return em.getVarExp(e.loc, sc);
         }
-
-        VarDeclaration v = s.isVarDeclaration();
-        if (v && (!v.type || !v.type.deco))
+        if (auto v = s.isVarDeclaration())
         {
-            if (v.inuse) // Bugzilla 9494
+            if (!v.type ||
+                !v.type.deco && v.inuse)
             {
-                e.error("circular reference to '%s'", v.toPrettyChars());
+                if (v.inuse)        // Bugzilla 9494
+                    e.error("circular reference to %s '%s'", v.kind(), v.toPrettyChars());
+                else
+                    e.error("forward reference to %s '%s'", v.kind(), v.toPrettyChars());
                 return new ErrorExp();
             }
-            if (v._scope)
+            if (v.type.ty == Terror)
+                return new ErrorExp();
+
+            if ((v.storage_class & STCmanifest) && v._init)
             {
-                v.semantic(v._scope);
-                s = v.toAlias(); // Need this if 'v' is a tuple variable
-                v = s.isVarDeclaration();
+                if (v.inuse)
+                {
+                    e.error("circular initialization of %s '%s'", v.kind(), v.toPrettyChars());
+                    return new ErrorExp();
+                }
+                checkAccess(e.loc, sc, null, v);
+                e = new VarExp(e.loc, v);
```

Comment:
Is there coding style? Actually such the replacement is actually used in `Dsymbol.resolve`, so I'm following it in here because I'm planning to mege the duplicated logic code into it.